### ISSUE TITLE
[rllib] Number of episodes in trainer.train() with "no_done_at_end : True" (?)

### ISSUE THREAD

**Issue by FruechteBini** [ray-project/ray#12709]
Hello guys,

i am currently working on a multi-agent project and i have a hard time understanding how the number of executed episodes is determined.

Info:

My train()-method is similar to the one on the [python api-documentation](https://docs.ray.io/en/master/rllib-training.html#basic-python-api):
```
trainer = DQNTrainer(config=config, env=my_env)
    for i in range(100):
		print("STARTING A NEW EPISODE:")
        result = trainer.train()
        print(pretty_print(result))
    trainer.save(os.path.join(saving_dir, "checkpoints"))
```
my config looks like this:
```
    rl_config = {
        "env": "StarWarsEnv",
        "env_config": {
            "num_agents": args.num_agents,
        },
        # "train_batch_size": 250,
        # "simple_optimizer": args.simple,
        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        # "num_sgd_iter": 1,
        "no_done_at_end": True,
        "num_workers": 0,
        "model": {"custom_model": CUSTOM_DENSE_KERAS},
        "evaluation_num_episodes": 10,
        "multiagent": {
            "policies": {
                "protagonist": (None, obs_space, act_space, {"gamma": 0.95}),
                "antagonist": (None, obs_space, act_space, {"gamma": 0.95})
            },
            "policy_mapping_fn": lambda agent_id:
            "antagonist"
            if agent_id.startswith("antagonist_")
            else "protagonist",
        },
        "framework": "torch" if args.torch else "tf",
        "callbacks": {
            "on_episode_start": tune.function(on_episode_start),
            "on_episode_end": tune.function(on_episode_end)
        }
    }
```
As you can see, i use custom callbacks to print the episode number at the beginning of an episode. 
```
def on_episode_start(info):
    global EPISODE_COUNTER
    print(f"---------episode {EPISODE_COUNTER}----------")
    EPISODE_COUNTER += 1
```

Problem:
The output looks like this, as you can see, the on_episode_start() is called and seems to be working, episodes are executed as expected (i checked it rendering and the reward is correct after each episode).
![image](https://user-images.githubusercontent.com/33086953/101625367-94ce2d00-3a1b-11eb-8e19-c6f8f61dfb59.png)

However, it seems like one iteration of trainer.train() executes multiple episodes itself?
As an example, in this run, the first trainer.train() call executed 185 episodes (this number varies each run):
--> i in the main-method train() is incremented after 185 episodes in this case.
![image](https://user-images.githubusercontent.com/33086953/101625229-694b4280-3a1b-11eb-8355-ae0883976930.png)

I did not find a parameter in the [trainer config](https://docs.ray.io/en/master/rllib-training.html#common-parameters) to set the number of episodes. But as on_episode_start() is called correctly the trainer obviously is handling episodes the correct way / does identify the episode is done?

I can only guess this has to do with one of the following scenarios:

- I had to set "no_done_at_end" to True, because my env is not resetting itself but rather respawns the agents. If this is the case, do i have to add anything?
- It has something to do with the batch size? Or max_steps set by the trainer and the e.g. 185 episodes in this case are equivalent to max_steps?

If this has been asked i am sorry, would appreciate every hint.

Thanks and greets

**Comment by sven1977** [comment]
It's totally normal to have more than one episode in each train iteration. This depends on:
1) Your environment and the number of timesteps that an episode lasts.
2) Your "horizon" config key, which will force-terminate the environment after that many timesteps (even if the episode is still ongoing).
3) Your "soft_horizon" config key: If True, RLlib will not try to reset the environment after the horizon is hit. It will, however, still reset the env, iff the env itself terminates (w/o hitting your horizon setting).
4) Your "no_done_at_end" config: If True, RLlib will not set the "done" column in the collected samples to True, but leave them at False, even though the episode ended (or the horizon was hit).
5) Your "rollout_fragment_length": This determines, how many steps (even across episodes) each worker rolls out before packing all the encountered data into a batch and sending it to the local worker for replay-buffering.
6) Your "batch_mode" setting (usually "truncate_episodes", but can be set to "complete_episodes"): This determines, whether a rollout is always exactly one episode (maybe this is the setting you would like to have?) or rather "rollout_fragment_length" steps long.

**Comment by FruechteBini** [comment]
Hey Sven, thanks for your reply and sorry for the late answer.
Thanks for your list - i played around with these parameters but still cant get it to work.
Lets say i want one iteration of trainer.train() to run exactly one episode.

I have set:
- all horizon related parameters are left to default
- my env has a max_steps variable itself, after which i set dones_dict['__all__'] = True, this seems to be working.

in the rl_config:
```
"no_done_at_end": True,
"rollout_fragment_length": 1,
"batch_mode": "complete_episodes",
"num_envs_per_worker": 1,
"num_workers": 0
```

i also tried to set the train_batch_size to 1.

The only quick fix i found was to do a sys.exit() in the on_episode_end() if an episode counter is reached. I guess for now its okay - however, if you have an idea i would appreciate it. But its not super important - thank you!

**Comment by jmugan** [comment]
I'm confused by this as well. I'd like each call to train to be an episode, or at least have some clear way to specify how many episodes to run.  Some environments don't have a clear end, but mine do, and so do things like games, and I think about it in those terms.

**Comment by vinzm** [comment]
> 6. Your "batch_mode" setting (usually "truncate_episodes", but can be set to "complete_episodes"): This determines, whether a rollout is always exactly one episode (maybe this is the setting you would like to have?) or rather "rollout_fragment_length" steps long.

I think it should be: 
This determines, whether a rollout is always *a whole number of episodes* (not necessarily just one, for example if rollout_fragment_length is much greater than the typical length, in steps, of an episode) or rather "rollout_fragment_length" steps long.

To be able to run a single episode for iteration, I had to set "train_batch_size" to 1, together with "batch_mode"="complete_episodes".

Hope this helps (even almost an year later...)
